libclamav/pe.c

```diff
@@ -2468,28 +2468,30 @@ static cl_error_t hash_imptbl(cli_ctx *ctx, unsigned char **digest, uint32_t *im
         }
     }
 
-    image = (struct pe_image_import_descriptor *)impdes;
-    while (left > sizeof(struct pe_image_import_descriptor) && image->Name != 0 && nimps < PE_MAXIMPORTS) {
+    memcpy(&image, impdes, sizeof(struct pe_image_import_descriptor));
+
+    while (left > sizeof(struct pe_image_import_descriptor) && image.Name != 0 && nimps < PE_MAXIMPORTS) {
         char *dllname = NULL;
 
         left -= sizeof(struct pe_image_import_descriptor);
         nimps++;
 
         /* Endian Conversion */
-        image->u.OriginalFirstThunk = EC32(image->u.OriginalFirstThunk);
-        image->TimeDateStamp        = EC32(image->TimeDateStamp);
-        image->ForwarderChain       = EC32(image->ForwarderChain);
-        image->Name                 = EC32(image->Name);
-        image->FirstThunk           = EC32(image->FirstThunk);
+        image.u.OriginalFirstThunk = EC32(image.u.OriginalFirstThunk);
+        image.TimeDateStamp        = EC32(image.TimeDateStamp);
+        image.ForwarderChain       = EC32(image.ForwarderChain);
+        image.Name                 = EC32(image.Name);
+        image.FirstThunk           = EC32(image.FirstThunk);
 
         /* DLL name acquisition */
-        offset = cli_rawaddr(image->Name, peinfo->sections, peinfo->nsections, &err, fsize, peinfo->hdr_size);
+        offset = cli_rawaddr(image.Name, peinfo->sections, peinfo->nsections, &err, fsize, peinfo->hdr_size);
         if (err || offset > fsize) {
             cli_dbgmsg("scan_pe: invalid rva for dll name\n");
             /* TODO: ignore or return? */
             /*
-              image++;
-              continue;
+            impdes += sizeof(struct pe_image_import_descriptor);
```

Comment:
It's not obvious (at first glance) what's going on here (i.e., what's the intended change to this pointer?).  It was originally defined as a `char *`, so this appears to be moving the pointer along a contiguous set of those structures.

Might be clearer if defined as `struct pe_image_import_descriptor *`, which would allow you to just do `impdes++` without specifying the size.

I also don't see the bounds check, but I'm not familiar with this structure.